backend/models/sequence.js

```diff
@@ -97,45 +191,92 @@ const getDefaultLegend = async (account, model) => {
 
 const Sequence = {};
 
-Sequence.getSequenceActivityDetail = async (account, model, activityId) => {
-	const activity = await db.findOne(account, activityCol(model), {"_id": utils.stringToUUID(activityId)});
+Sequence.createSequence = async (account, model, data) => {
+	const newSequence = {
+		"_id": utils.generateUUID(),
+		"customSequence": true,
+		"frames": []
+	};
+
+	if (!data || !data.name || !data.frames) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	} else {
+		newSequence.name = data.name;
+	}
 
-	if (!activity) {
-		throw responseCodes.ACTIVITY_NOT_FOUND;
+	if (data.rev_id) {
+		const history = await History.getHistory(account, model, undefined, data.rev_id, {_id: 1});
+		newSequence.rev_id = history._id;
 	}
 
-	return cleanActivityDetail(activity);
+	newSequence.frames = await handleFrames(account, model, newSequence._id, data.frames);
+
+	newSequence.startDate = new Date((newSequence.frames[0] || {}).dateTime);
+	newSequence.endDate = new Date((newSequence.frames[newSequence.frames.length - 1] || {}).dateTime);
+
+	if (!sequenceSchema.isValidSync(newSequence, { strict: true })) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	await db.insert(account, sequenceCol(model), newSequence);
+
+	return { _id: utils.uuidToString(newSequence._id) };
 };
 
-Sequence.getSequenceActivities = async (account, model, sequenceId) => {
-	return FileRef.getSequenceActivitiesFile(account, model, utils.uuidToString(sequenceId));
+Sequence.deleteSequence = async (account, model, sequenceId) => {
+	await Sequence.sequenceExists(account, model, sequenceId);
+	const { result } = await db.remove(account, sequenceCol(model), {
+		_id: utils.stringToUUID(sequenceId),
+		customSequence: true
+	});
+
+	if (result.n === 0) {
+		throw responseCodes.SEQUENCE_READ_ONLY;
+	}
+};
+
+Sequence.getSequenceById = async (account, model, sequenceId, projection = {}, noClean = true) => {
+	const sequence = await db.findOne(account, sequenceCol(model), { _id: utils.stringToUUID(sequenceId)}, projection);
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+
+	return noClean ? sequence : cleanSequence(account, model, sequence);
+};
+
+Sequence.sequenceExists = async (account, model, sequenceId) => {
+	if(!(await Sequence.getSequenceById(account, model, utils.stringToUUID(sequenceId), {_id: 1}))) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
 };
 
 Sequence.getSequenceState = async (account, model, stateId) => {
 	return FileRef.getSequenceStateFile(account, model, stateId);
 };
 
 Sequence.getList = async (account, model, branch, revision, cleanResponse = false) => {
-	const history = await History.getHistory(account, model, branch, revision);
+	let submodelBranch;
+	let sequencesQuery = {};
 
-	if(!history) {
-		throw responseCodes.INVALID_TAG_NAME;
+	if (branch || revision) {
+		const history = await History.getHistory(account, model, branch, revision, {_id: 1});
+
+		submodelBranch = "master";
```

Comment:
@carmenfan @mosment3DRepo There is a slight difference and it depends what data the user should be able to query. At the moment, we're allowing users to get ALL federation and sub-model sequences.

If we make sub-model revision branch to always be "master" then they will only ever get the sequences for the latest revision + sequences with unspecified revision in the sub-models. For models and the federation, we retrieve all sequences if no revision is specified.

With this change it won't retrieve all sub-model sequences.

Reply:
@carmenfan @mosment3DRepo In the case where the user does specify a specific revision, for federations it will retrieve sequences from that revision + sequences with unspecified revision and the latest "master" sequences from sub-models.

Reply:
in that case let's leave it as is.